### ISSUE TITLE
Wrong Content-Type value when returning Typed JSON from handler

### ISSUE THREAD

**Issue by Mackiovello** [Starcounter/Home#76]
With the following handler:
```cs
Handle.GET("/GetPerson", () => 
{
    return new PersonPage();
});
```

The response header Content-Type is set to `text/html` when it should be `application/json`, at least according to [the documentation](https://docs.starcounter.io/guides/typed-json/responding-with-json/)

@alemoi @chrhol

**Comment by chrhol** [comment]
Yes, there seem to be some inconsistency here, both in how it's returned and whats described in the documentation. 

There are some built-in functionality in and around TypedJSON that will try to convert a TypedJSON instance to a representation for what the client asks for. If I enter an url in a browser (I assumed that was what you did when testing the above) the client will ask for `text/html`. 

The default behaviour of any TypedJSON instance is to simply return itself as json, not caring about the mimetype at all, but it doesn't set the outgoing mimetype in the response.

I'm not sure how to deal with this, or what the desired behaviour is.

- Is it ok to respond with a mimetype the client didn't ask for or that is not included in the `Accept` header? 
- Do we still need this automatic "magic" conversion of TypedJSON (i.e. from `IResource` and `IResponseConverter`) or should we instead handle this more explicitly, and/or using middleware?

I'm pinging @warpech, @per-samuelsson @miyconst as well. I think we need to decide on how we want it to work and not just do a quick fix by setting correct mimetype and/or updating the docs.

**Comment by miyconst** [comment]
>Do we still need this automatic "magic" conversion of TypedJSON (i.e. from `IResource` and `IResponseConverter`) or should we instead handle this more explicitly, and/or using middleware?

I would rather remove this feature and provide single default behavior for every type accepted as a response. If something extra is required then middleware filters should be used.

Let's see what @warpech thinks.

**Comment by warpech** [comment]
The browser request is:

```
Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
```

Which means, give me in that order or prority:

1. HTML representation of what I am requesting
2. XHTML
3. XML
4. WebP (in case that is an image, you just might want to know I can render WebP's)
5. Any other representation

So it is correct that we respond with HTML if we think that a JSON has a HTML representation.

But we should never send JSON content with `text/html` mime type. We should only send something with `text/html` mime type if that is actually HTML content.

So I am not saying "remove any feature". I am just saying remove sending wrong mime type header if what you are sending to the client is serialized JSON.

**Comment by chrhol** [comment]
> So I am not saying "remove any feature". 

That was my question :) Do we really need this feature. 

> I just think that it is a bug somewhere that we are sending serialized JSON with incorrect Content-Type header. Maybe we incorrectly take the value of the response header from the request?

Yes, I just wanted to make sure that it was correct to do so, and not just a quick and dirty hack. 

I can easily change so that the correct mimetype is used on the outgoing response for this case, I just wasn't sure that this was the way to do it or if there was some underlying magic that was the root of the problem. 

And I think we should at least consider going through this functionality and see if a) it's used at all, and b) do we need it?

**Comment by per-samuelsson** [comment]
I remember I reflected over this somewhere at the time I wrote this:

https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers

It's probably burried somewhere in an issue somewhere out there already. 😄

**Comment by chrhol** [comment]
> I can easily change so that the correct mimetype is used on the outgoing response

Well, I might have been a bit optimistic here. The current implementation of  `IResource` and `IResponseConverter` interfaces in TypedJSON is confusing to say the least. The implementations call eachother from different places. Some parts care about the mimetype while other just ignores it. And we never check for any other mimetype than the first one.

I think I will spend a little time trying to enhance the implementations a bit.

> I remember I reflected over this somewhere at the time I wrote this:

Thank you for the link. Maybe we should put that document somewhere more visible?

**Comment by chrhol** [comment]
Fix pushed to v2.3 and develop now.

**Comment by chrhol** [comment]
I will close this as done.
